idx/memory/memory.go

```diff
@@ -269,6 +374,101 @@ func (m *MemoryIdx) GetPath(orgId int, path string) []idx.Archive {
 	return archives
 }
 
+func (m *MemoryIdx) Tag(orgId int, tag string, from int64) map[string]uint32 {
+	if !tagSupport {
+		log.Warn("memory-idx: received tag query, but tag support is disabled")
+		return nil
+	}
+
+	m.RLock()
+	defer m.RUnlock()
+
+	var tags TagIndex
+	var ok bool
+	if tags, ok = m.Tags[orgId]; !ok {
+		return nil
+	}
+
+	result := make(map[string]uint32)
+
+	for value, ids := range tags[tag] {
+		valueCnt := uint32(0)
+		for id := range ids {
+			var def *idx.Archive
+			var ok bool
+			if def, ok = m.DefById[id.String()]; !ok {
+				// should never happen because every ID that is in the tag index
+				// must be present in the byId lookup table
+				CorruptIndex.Inc()
+				log.Error(3, "memory-idx: ID %s is in tag index but not in the byId lookup table", id.String())
+				continue
+			}
+
+			if def.LastUpdate < from {
+				continue
+			}
+
+			valueCnt++
+		}
+		if valueCnt > 0 {
+			result[value] = valueCnt
+		}
+	}
+
+	return result
+}
+
+func (m *MemoryIdx) TagList(orgId int) []string {
+	if !tagSupport {
+		log.Warn("memory-idx: received tag query, but tag support is disabled")
+		return nil
+	}
+
+	m.RLock()
+	defer m.RUnlock()
+
+	var tags TagIndex
+	var ok bool
+	if tags, ok = m.Tags[orgId]; !ok {
+		return nil
+	}
+
+	results := make([]string, 0, len(tags))
+
+	for k := range tags {
+		results = append(results, k)
+	}
+
+	return results
+}
+
+func (m *MemoryIdx) IdsByTagExpressions(orgId int, expressions []string, from int64) (map[idx.MetricID]struct{}, error) {
+	query, err := NewTagQuery(expressions, from)
+	if err != nil {
+		return nil, err
+	}
+
+	seriesMap := m.IdsByTagQuery(orgId, query)
+	return seriesMap, nil
+}
+
+func (m *MemoryIdx) IdsByTagQuery(orgId int, query TagQuery) TagIDs {
```

Comment:
is it needed for this to be a separate function, rather than just be embedded in the function above? it seems this would actually slightly improve the unit tests as well as you can then get the validation of the tag query and the execution of this in 1 shot.

Reply:
either way this should not be an exported function

Reply:
The main reason why i separated them was because I wanted to be able to test the building of a query and the execution of a query separately. Will make it non-exported